How to write a new Foreman Ansible module post (#1692)
theforeman/theforeman.org: _posts/2020-10-05-how-to-write-a-new-foreman-ansible-module-in-20-lines-of-code.md

---
layout: post
title: How to write a new Foreman Ansible module in 20 lines of code
date: 2020-10-05 11:29:55
author: Evgeni Golov
tags:
- foreman
---

Some time ago, we were asked ([twice](https://bugzilla.redhat.com/show_bug.cgi?id=1856401)) to [add a module for managing HTTP proxies in Foreman](https://github.com/theforeman/foreman-ansible-modules/issues/867) to the Foreman Ansible Modules collection. Given the limited scope of the module, it's a perfect fit for a guide how to write new modules utilizing the abstraction we have developed in Foreman Ansible Modules!

> <!--more-->

Reviewer: ??

Author: this is a marker for Jekyll to add the first part on the main https://theforeman.org/blog/ page and the rest after the reader opens the real post.

    pass
```

Now we need to add the other fields the API supports as parameters. We'll mark the `name` parameter as `required` because that's the one we'll be using for finding existing proxies -- the API has it as `optional` because it uses the ID of the proxy as the primary identification, but nobody wants to write static ids in Ansible playbooks. The `url` parameter is only required for *new* proxies, so we'll mark it as `optional` in the module parameters and document the requirement for creation. As our modules already have `username` and `password` parameters for authenticating against Foreman, we'll have to name the parameters for those fields differently -- `proxy_` seems like a good prefix to add here. And last but not least, `password` parameters should be marked as `no_log=True` so that Ansible skips them when printing debugging information about the module invocation:

```python
    module = ForemanHttpProxyModule(
        foreman_spec=dict(
            name=dict(required=True),
            url=dict(),
            proxy_username=dict(flat_name='username'),
            proxy_password=dict(no_log=True, flat_name='password'),
        ),
    )
```

> `flat_name` is telling our abstraction layer that the field name in the API differs from the Ansible parameter name.

Reviewer: Again a reason to mention the documentation? There are a lot of additional options. Also we support everything argument_spec does. Maybe we even state that foremen_spec is a replacement.

The most interesting part of the `DOCUMENTATION` stanza is the `extends_documentation_fragment` section. As we have several parameters coming from our base class, we also have documentation fragments that document these parameters:

```yaml
extends_documentation_fragment:
  - theforeman.foreman.foreman
  - theforeman.foreman.foreman.entity_state
  - theforeman.foreman.foreman.taxonomy
```

`theforeman.foreman.foreman` adds the most basic parameters for the API connection, `theforeman.foreman.foreman.entity_state` adds the `state` parameter for our entity and finally `theforeman.foreman.foreman.taxonomy` adds `organizations` and `locations` parameters to describe the taxonomy.

The rest of the `DOCUMENTATION` section is basic [Ansible module documentation](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html), so we'll omit it here.

## Tests!

Writing tests is tedious and annoying, but at the same time, a lot of bugs can be prevented if you have tests. Or could have been prevented, if there would have been tests ;-)

> In Foreman Ansible Modules, we mostly do integration tests -- unit testing modules is not really feasible. Those integration tests consist of an Ansible playbook that executes the module multiple times and analyses the results and a set of recorded API calls that can be used to replay the tests without having a full Foreman running, which saves a lot of resources when running the tests.

Reviewer: It adds to the resilience of the tests, IMHO.

Reviewer: Also as i recently figured, it allows to run the tests in parallel.

Now the main playbook (`http_proxy.yml`) is a bit longer, so we won't paste it verbatim here, but discuss the individual sections.

First, it has a play with `hosts: localhost` to prepare the test environment: setup organizations, locations and whatever else might be needed.

This is followed by a long play with `hosts: tests` that contains the actual test steps.
Those steps follow a pattern: do something and expect a change to happen (`expected_change: true`), do the same thing again and expect no change to happen (`expected_change: false`). This ensures idempotency of the module.

The "things to do" again follow a pattern: create a proxy with minimal information, update a proxy, delete a proxy. This ensures the common workflows of a user are covered.

> And last, it has is another play with `hosts: localhost` to tear down anything that has been prepared for this specific test to run.

Reviewer: ```suggestion
At last, it has another play with `hosts: localhost` to tear down anything that has been prepared for this specific test to run.
```

The differentiation between `hosts: localhost` and `hosts: tests` allows us to execute the `localhost` steps only when we're recording a new set of API interactions and the `tests` steps always, as there is no need for the setup and teardown when running the tests against the recorded data.

The last missing piece between us and working tests is a symlink. To keep the recorded API data clean of the big apidoc JSON and also allow easier updates when the apidoc changes, we keep copies of the JSON in `tests/fixtures/apidoc`. However, we keep only `foreman.json`, `katello.json` and `luna.json` as real files while all the others are symlinks. As the HTTP Proxy is a Foreman feature, we will create a symlink from `http_proxy.json` to `foreman.json`. Would it have been a Katello feature, the link would have been to `katello.json`. And for any other plugin, it's `luna.json` as that's essentially "Foreman with all the nice plugins".

> Having all pieces in place (and a running Foreman instance somewhere), we can run `make record_http_proxy` which it will execute our playbook and record the API interactions in `tests/test_playbooks/fixtures`.

Reviewer: Ideas:
"Don't forget to check those fixtures in when submitting a PR."
"You can rerecord at any time."
"Use `make livetest_http_proxy` to test your playbook locally."
"The recorded fixtures are verified with `make test_http_proxy`."
